File: cpp/src/kdtree.cpp
#include "kdtree.hpp"
#include <cmath>
#include <algorithm>

namespace ds {
KDTree::KDTree() : root(nullptr) {}

KDTree::~KDTree() {
    auto freeKD = [](KDNode* n) {
        if (!n) return;
        freeKD(n->left);
        freeKD(n->right);
        delete n;
    };
    freeKD(root);
}

long long KDTree::dist(int r1, int g1, int b1, int r2, int g2, int b2) {
    long long dr = r1 - r2, dg = g1 - g2, db = b1 - b2;
    return dr * dr + dg * dg + db * db;
}

KDNode* KDTree::build(std::vector<std::pair<std::array<int,3>, std::string>>& points, int depth) {
    if (points.empty()) return nullptr;
    int axis = depth % 3;
    std::nth_element(points.begin(), points.begin() + points.size() / 2, points.end(),
        [axis](const auto& a, const auto& b) { return a.first[axis] < b.first[axis]; });
    auto mid = points.begin() + points.size() / 2;
    KDNode* n = new KDNode((*mid).first[0], (*mid).first[1], (*mid).first[2], (*mid).second);
    std::vector<std::pair<std::array<int,3>, std::string>> left(points.begin(), mid),
        right(mid + 1, points.end());
    n->left = build(left, depth + 1);
    n->right = build(right, depth + 1);
    return n;
}

void KDTree::buildTree(const std::vector<std::pair<std::array<int,3>, std::string>>& colors) {
    auto pts = colors;
    root = build(pts, 0);
}

void KDTree::nearest(KDNode* node, int r, int g, int b, int depth, KDNode*& best, long long& bestDist) {
    if (!node) return;
    long long d = dist(r, g, b, node->r, node->g, node->b);
    if (d < bestDist) { bestDist = d; best = node; }
    int axis = depth % 3;
    int val = (axis == 0) ? r : (axis == 1) ? g : b;
    int nodeVal = (axis == 0) ? node->r : (axis == 1) ? node->g : node->b;
    KDNode* near = (val < nodeVal) ? node->left : node->right;
    KDNode* far = (val < nodeVal) ? node->right : node->left;
    nearest(near, r, g, b, depth + 1, best, bestDist);
    long long planeDist = (val - nodeVal) * (long long)(val - nodeVal);
    if (planeDist < bestDist) nearest(far, r, g, b, depth + 1, best, bestDist);
}

std::string KDTree::findNearest(int r, int g, int b) {
    if (!root) return "";
    KDNode* best = root;
    long long bestDist = dist(r, g, b, root->r, root->g, root->b);
    nearest(root, r, g, b, 0, best, bestDist);
    return best ? best->name : "";
}
} // namespace ds

File: universe/src/modules/toolkit/lib/kdTree.ts
/**
 * 三维 KD-Tree（RGB 空间），用于最近邻与中国色匹配
 */

export interface KDPoint<T> {
  coords: [number, number, number];
  data: T;
}

export class KDTreeNode<T> {
  constructor(
    public point: KDPoint<T>,
    public axis: number,
    public left: KDTreeNode<T> | null = null,
    public right: KDTreeNode<T> | null = null
  ) {}
}

function distSq(a: [number, number, number], b: [number, number, number]): number {
  const dx = a[0] - b[0];
  const dy = a[1] - b[1];
  const dz = a[2] - b[2];
  return dx * dx + dy * dy + dz * dz;
}

function build<T>(points: KDPoint<T>[], depth: number): KDTreeNode<T> | null {
  if (points.length === 0) return null;
  const axis = depth % 3;
  const sorted = [...points].sort((p, q) => p.coords[axis] - q.coords[axis]);
  const mid = Math.floor(sorted.length / 2);
  const median = sorted[mid]!;
  const left = build(sorted.slice(0, mid), depth + 1);
  const right = build(sorted.slice(mid + 1), depth + 1);
  return new KDTreeNode(median, axis, left, right);
}

/** K 近邻用：最大堆存 (distSq, point)，堆顶为当前 K 个里最远的 */
interface Neighbor<T> {
  d: number;
  p: KDPoint<T>;
}

export class KDTree<T> {
  private root: KDTreeNode<T> | null;

  constructor(points: KDPoint<T>[]) {
    this.root = points.length ? build(points, 0) : null;
  }

  nearest(target: [number, number, number]): KDPoint<T> | null {
    if (!this.root) return null;
    let best: KDPoint<T> | null = null;
    let bestD = Infinity;

    const search = (node: KDTreeNode<T> | null) => {
      if (!node) return;
      const d = distSq(target, node.point.coords);
      if (d < bestD) {
        bestD = d;
        best = node.point;
      }
      const ax = node.axis;
      const diff = target[ax] - node.point.coords[ax];
      const first = diff <= 0 ? node.left : node.right;
      const second = diff <= 0 ? node.right : node.left;
      search(first);
      if (diff * diff < bestD) search(second);
    };

    search(this.root);
    return best;
  }

  /**
   * K 最近邻：最大堆维护当前 K 个候选中「最远」的一个，便于剪枝
   */
  nearestK(target: [number, number, number], k: number): KDPoint<T>[] {
    if (!this.root || k <= 0) return [];
    const heap: Neighbor<T>[] = [];

    const sortMaxRoot = () => heap.sort((a, b) => b.d - a.d);

    const tryAdd = (p: KDPoint<T>) => {
      const d = distSq(target, p.coords);
      if (heap.length < k) {
        heap.push({ d, p });
        sortMaxRoot();
        return;
      }
      if (d >= heap[0]!.d) return;
      heap[0] = { d, p };
      sortMaxRoot();
    };

    const worstDist = () => (heap.length < k ? Infinity : heap[0]!.d);

    const search = (node: KDTreeNode<T> | null) => {
      if (!node) return;
      tryAdd(node.point);
      const ax = node.axis;
      const diff = target[ax] - node.point.coords[ax];
      const first = diff <= 0 ? node.left : node.right;
      const second = diff <= 0 ? node.right : node.left;
      search(first);
      const w = worstDist();
      if (diff * diff < w || heap.length < k) search(second);
    };

    search(this.root);
    return heap.sort((a, b) => a.d - b.d).map((n) => n.p);
  }
}
